apps/champions/lib/champions/battle/simulator.ex

```diff
@@ -63,187 +64,459 @@ defmodule Champions.Battle.Simulator do
+          Logger.info("Process step #{step} for effect cast by #{format_unit_name(effect.caster)}")
+
+          process_step_for_effect(effect, current_state)
+        end)
+
+      Logger.info("Step #{step} finished: #{inspect(format_step_state(new_state))}")
+
       remove_dead_units(new_state)
       |> check_winner(step)
     end)
   end
 
-  defp process_step(unit, initial_step_state, current_state) do
+  # Removes dead units from the battle state.
+  defp remove_dead_units(state) do
+    new_units =
+      Enum.reduce(state.units, %{}, fn {unit_id, unit}, units ->
+        if unit.health > 0 do
+          Map.put(units, unit_id, unit)
+        else
+          Logger.info("Unit #{format_unit_name(unit)} died.")
+          units
+        end
+      end)
+
+    Map.put(state, :units, new_units)
+  end
+
+  # Check if the battle has ended.
+  # Battle can end if all unit of a team are dead, or if the maximum step amount has been reached.
+  defp check_winner(state, step) do
+    winner =
+      cond do
+        Enum.empty?(state.units) -> :tie
+        Enum.all?(state.units, fn {_id, unit} -> unit.team == 2 end) -> :team_2
+        Enum.all?(state.units, fn {_id, unit} -> unit.team == 1 end) -> :team_1
+        true -> :none
+      end
+
+    case winner do
+      :none ->
+        if step == @maximum_steps do
+          Logger.info("Battle timeout.")
+          {:halt, :timeout}
+        else
+          {:cont, state}
+        end
+
+      result ->
+        Logger.info("Battle ended. Result: #{result}.")
+        {:halt, result}
+    end
+  end
+
+  # Calculate the new state of the battle after a step passes for a unit.
+  # Updates cooldowns, casts skills and reduces self-affecting modifier durations.
+  defp process_step_for_unit(unit, current_state, initial_step_state) do
     new_state =
       cond do
-        not can_attack(unit) ->
+        not can_attack(unit, initial_step_state) ->
+          Logger.info("Unit #{format_unit_name(unit)} cannot attack")
           current_state
 
         can_cast_ultimate_skill(unit) ->
-          unit
-          |> cast_skill(unit.ultimate_skill, initial_step_state, current_state)
-          |> put_in([unit.id, :energy], 0)
+          Logger.info("Unit #{format_unit_name(unit)} casting Ultimate skill")
+
+          current_state
+          |> Map.put(:skills_being_cast, [unit.ultimate_skill | current_state.skills_being_cast])
+          |> put_in([:units, unit.id, :energy], 0)
 
         can_cast_basic_skill(unit) ->
-          new_state = cast_skill(unit, unit.basic_skill, initial_step_state, current_state)
+          Logger.info("Unit #{format_unit_name(unit)} casting basic skill")
 
-          new_state
+          current_state
+          |> Map.put(:skills_being_cast, [unit.basic_skill | current_state.skills_being_cast])
           |> put_in(
-            [unit.id, :basic_skill, :remaining_cooldown],
+            [:units, unit.id, :basic_skill, :remaining_cooldown],
```

Comment:
Won't this operation always remain the same? is `unit.basic_skill.base_cooldown` being updated somewhere? you can also replace this put in an update_in

Reply:
right!